### PR TITLE
basic cleanbots refactor and new janitor skillchip

### PR DESCRIPTION
**Pull request by Ben10Omintrix**
## About The Pull Request
this pr refacotrs cleanbots into basic bots. also adds a new skillchip for janitrs. this skillchip will allow janitors to communicate with cleanbots and order them around, like pointing at something and telling them to clean it. also now the cleanbot has an inbuilt mop which it will use to smack mice and cockroaches

## Why It's Good For The Game
refactors cleanbots into basic bots and fixes them getting stuck sometimes while patrolling. also janitors being able to order them around can make them a bit more useful as tools for the janitor

## Changelog
:cl:
refactor: cleanbots are refactored into basic bots. please report all bugs 
fix: fixes cleanbots getting stuck sometimes while patrolling
add: janitors get a new skillchip which allow them to communicate with cleanbots
/:cl:

### CODE REVIEW

**Comment by Jolly-66** [comment]
Don't forget to replace the `/:cl` with 🆑, the actual emoji

**Comment by DrAmazing343** [comment]
ANOTHER BASED BEN10 PR! Always love to see more Janitor content.

**Comment by san7890** [comment]
> Don't forget to replace the `/:cl:` with 🆑, the actual emoji

no this doesn't work with the changelog processor as of present. `/:cl:` does though.

File: code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+	icon_state = "cleanbot0"
+	pass_flags = PASSMOB | PASSFLAPS
+	density = FALSE
+	anchored = FALSE
+	health = 25
+	maxHealth = 25
+
+	maints_access_required = list(ACCESS_ROBOTICS, ACCESS_JANITOR)
+	radio_key = /obj/item/encryptionkey/headset_service
+	radio_channel = RADIO_CHANNEL_SERVICE //Service //true
```
Comment (Jacquerel) [inline code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm:16]:
```suggestion
	radio_channel = RADIO_CHANNEL_SERVICE
```

File: code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+	hackables = "cleaning software"
+	additional_access = /datum/id_trim/job/janitor
+	greyscale_config = /datum/greyscale_config/buckets_cleanbot
+	possessed_message = "You are a cleanbot! Clean the station to the best of your ability!"
+	ai_controller = /datum/ai_controller/basic_controller/bot/cleanbot
+	///the bucket used to build us.
+	var/obj/item/reagent_containers/cup/bucket/build_bucket
+	///Flags indicating what kind of cleanables we should scan for to set as our target to clean.
+	var/janitor_mode_flags = CLEANBOT_CLEAN_BLOOD
+	//Selections: CLEANBOT_CLEAN_BLOOD | CLEANBOT_CLEAN_TRASH | CLEANBOT_CLEAN_PESTS | CLEANBOT_CLEAN_DRAWINGS
```
Comment (Jacquerel) [inline code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm:27]:
```suggestion
	///Flags indicating what kind of cleanables we should scan for to set as our target to clean.
	///Options: CLEANBOT_CLEAN_BLOOD | CLEANBOT_CLEAN_TRASH | CLEANBOT_CLEAN_PESTS | CLEANBOT_CLEAN_DRAWINGS
	var/janitor_mode_flags = CLEANBOT_CLEAN_BLOOD
```

File: code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+
+/mob/living/basic/bot/cleanbot/Initialize(mapload)
+	. = ..()
+
+	generate_ai_keys()
+	AddComponent(/datum/component/obeys_commands, pet_commands)
+	AddComponent(/datum/component/cleaner, 1 SECONDS, \
+		pre_clean_callback = CALLBACK(src, PROC_REF(update_bot_mode), BOT_CLEANING), \
+		on_cleaned_callback = CALLBACK(src, PROC_REF(update_bot_mode), BOT_IDLE), \
+	)
```
Comment (Jacquerel) [inline code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm:135]:
```suggestion
	AddComponent(/datum/component/cleaner, \
		base_cleaning_duration = 1 SECONDS, \
		pre_clean_callback = CALLBACK(src, PROC_REF(update_bot_mode), BOT_CLEANING), \
		on_cleaned_callback = CALLBACK(src, PROC_REF(update_bot_mode), BOT_IDLE), \
	)
```

File: code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+	SIGNAL_HANDLER
+	if(!istype(used_item, /obj/item/knife) || user.combat_mode)
+		return
+	INVOKE_ASYNC(src, PROC_REF(attach_knife), user, used_item)
+	return COMPONENT_NO_AFTERATTACK
+
+/mob/living/basic/bot/cleanbot/proc/attach_knife(mob/living/user, obj/item/used_item)
+	balloon_alert(user, "attaching knife...")
+	if(!do_after(user, 2.5 SECONDS, target = src))
+		balloon_alert(user, "need to stay still")
```
Comment (Jacquerel) [inline code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm:270]:
```suggestion
		balloon_alert(user, "need to stay still!")
```
Although frankly you probably don't need this one.

File: code/modules/mob/living/basic/bots/cleanbot/cleanbot_abilities.dm [hunk 5]
```diff
@@ -0,0 +1,36 @@
+/datum/action/cooldown/mob_cooldown/bot
+	background_icon_state = "bg_revenant"
+	overlay_icon_state = "bg_revenant_border"
```
Comment (Jacquerel) [inline code/modules/mob/living/basic/bots/cleanbot/cleanbot_abilities.dm:3]: Why revenant over tech or tech_blue?
No judgement if you think it looks better, just would have figured those were appropriate for a bot.

File: code/modules/mob/living/basic/bots/cleanbot/cleanbot_ai.dm [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+	if(!iscarbon(controller.blackboard[target_key]))
+		controller.clear_blackboard_key(target_key)
+		return
+	var/list/speech_list = controller.blackboard[BB_CLEANBOT_EMAGGED_PHRASES]
+	if(!length(speech_list))
+		return
+	var/mob/living/living_pawn = controller.pawn
+	living_pawn.say(pick(controller.blackboard[BB_CLEANBOT_EMAGGED_PHRASES]), forced = "ai controller")
+	controller.clear_blackboard_key(target_key)
+/datum/ai_planning_subtree/use_mob_ability/foam_area
```
Comment (Jacquerel) [inline code/modules/mob/living/basic/bots/cleanbot/cleanbot_ai.dm:137]:
```suggestion
	controller.clear_blackboard_key(target_key)

/datum/ai_planning_subtree/use_mob_ability/foam_area
```

File: code/modules/mob/living/simple_animal/bot/construction.dm [hunk 7]
```diff
@@ -75,16 +75,17 @@
 	return ..()
 
 
-/obj/item/bot_assembly/cleanbot/attackby(obj/item/W, mob/user, params)
+/obj/item/bot_assembly/cleanbot/attackby(obj/item/item_attached, mob/user, params)
 	..()
-	if(istype(W, /obj/item/bodypart/arm/left/robot) || istype(W, /obj/item/bodypart/arm/right/robot))
-		if(!can_finish_build(W, user))
+	if(istype(item_attached, /obj/item/bodypart/arm/left/robot) || istype(item_attached, /obj/item/bodypart/arm/right/robot))
```
Comment (Jacquerel) [inline code/modules/mob/living/simple_animal/bot/construction.dm:80]: You didn't write this but we can probably early return here

**Review by Jacquerel — CHANGES_REQUESTED** [review]

**Comment by Ben10Omintrix** [comment]
> Stupid question, would this fix the tile bots too? They seem to not want to tile or even repair floors.

no this is just cleanbots. ill fix tile bots when its time to refactor them

**Review by san7890 — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+		JOB_STATION_ENGINEER = "Engineer",
+		JOB_ATMOSPHERIC_TECHNICIAN = "Technician",
+		JOB_CHIEF_MEDICAL_OFFICER = "C.M.O.",
+		JOB_MEDICAL_DOCTOR = "M.D.",
+		JOB_CHEMIST = "Pharm.D.",
+		JOB_RESEARCH_DIRECTOR = "Ph.D.",
+		JOB_ROBOTICIST = "M.S.",
+		JOB_SCIENTIST = "B.S.",
+		JOB_GENETICIST = "Gene B.S.",
+		JOB_LAWYER = "Esq.",
```
[the remark this answers is not in the document]
Comment (Ben10Omintrix) [reply code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm:65]: i sorted them based on departments

[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm:65]: separate them by line breaks then

File: code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+	var/obj/item/mop/our_mop
+	///list of our officer titles
+	var/static/list/officers_titles = list(
+		JOB_CAPTAIN,
+		JOB_HEAD_OF_PERSONNEL,
+		JOB_HEAD_OF_SECURITY,
+		JOB_RESEARCH_DIRECTOR,
+	)
+	///job titles we can get
+	var/static/list/job_titles = list(
```
Comment (Mothblocks) [inline code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm:48]: feels like we should unit test that every job is in here

File: code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+	/// if we have all the top titles, grant achievements to living mobs that gaze upon our cleanbot god
+	var/ascended = FALSE
+	///List of all stolen names the cleanbot currently has.
+	var/list/stolen_valor = list()
+	///Currently attached weapon, usually a knife.
+	var/obj/item/weapon
+	///our mop item
+	var/obj/item/mop/our_mop
+	///list of our officer titles
+	var/static/list/officers_titles = list(
```
Comment (Mothblocks) [inline code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm:41]: i don't understand what this does or why who's in it is in it

File: code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+	balloon_alert(user, "attaching knife...")
+	if(!do_after(user, 2.5 SECONDS, target = src))
+		return
+	deputize(used_item, user)
+
+/mob/living/basic/bot/cleanbot/proc/deputize(obj/item/knife, mob/user)
+	if(!in_range(src, user) || !user.transferItemToLoc(knife, src))
+		balloon_alert(user, "couldn't attach!")
+		return FALSE
+	balloon_alert(user, "attached!")
```
Comment (Mothblocks) [inline code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm:283]:
```suggestion
	balloon_alert(user, "attached")
```

"!" == failure

File: code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm (hunk 9 above)
Comment (Ben10Omintrix) [reply code/modules/mob/living/basic/bots/cleanbot/cleanbot.dm:48]: this doesnt have all jobs. just a select few

**Comment by MTandi** [comment]
@Ben10Omintrix After this PR, they have prioritization issues, same as before #71159

**Pull request merged**